core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -575,4 +600,12 @@ public static StreamInput wrap(byte[] bytes) {
     public static StreamInput wrap(byte[] bytes, int offset, int length) {
         return new InputStreamStreamInput(new ByteArrayInputStream(bytes, offset, length));
     }
+
+    /**
+     * FunctionalInterface for methods that read an object from the stream.
+     */
+    @FunctionalInterface
+    public static interface ObjectReader<T> {
+        public T read(StreamInput in) throws IOException;
+    }
```

Comment:
could this part of the PR be moved to another PR? Sorry I know this is a bit painful but since you asked for a careful review this would help me ensure everything is correct